libc/src/__support/threads/linux/raw_mutex.h

```diff
@@ -0,0 +1,116 @@
+//===--- Implementation of a Linux RawMutex class ---------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_LIBC_SRC___SUPPORT_THREADS_LINUX_RAW_MUTEX_H
+#define LLVM_LIBC_SRC___SUPPORT_THREADS_LINUX_RAW_MUTEX_H
+
+#include "futex_word.h"
+#include "src/__support/CPP/optional.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/attributes.h"
+#include "src/__support/threads/linux/futex_utils.h"
+#include "src/__support/threads/sleep.h"
+#include "src/__support/time/linux/abs_timeout.h"
+#ifdef LIBC_COPT_TIMEOUT_ENSURE_MONOTONICITY
+#include "src/__support/time/linux/monotonicity.h"
+#endif
+
+#ifndef LIBC_COPT_RAW_MUTEX_DEFAULT_SPIN_COUNT
+#define LIBC_COPT_RAW_MUTEX_DEFAULT_SPIN_COUNT 100
+#endif
+
+namespace LIBC_NAMESPACE {
+namespace internal {
+// Lock is a simple timable lock for internal usage.
+// This is separated from Mutex because this one does not need to consider
+// robustness and reentrancy. Also, this one has spin optimization for shorter
+// critical sections.
+class RawMutex {
+protected:
+  Futex futex;
+  LIBC_INLINE_VAR static constexpr FutexWordType UNLOCKED = 0b00;
+  LIBC_INLINE_VAR static constexpr FutexWordType LOCKED = 0b01;
+  LIBC_INLINE_VAR static constexpr FutexWordType CONTENTED = 0b10;
+
+  LIBC_INLINE FutexWordType spin(uint_fast32_t spin_count) {
+    FutexWordType result;
+    for (;;) {
+      result = futex.load(cpp::MemoryOrder::RELAXED);
+      // spin until one of the following conditions is met:
+      // - the mutex is unlocked
+      // - the mutex is contented
+      // - the spin count reaches 0
+      if (result != LOCKED || spin_count == 0)
+        return result;
+      // Pause the pipeline to avoid extraneous memory operations due to
+      // speculation.
+      sleep_briefly();
```

Comment:
Modern mutex implementations often do a few spins before waiting. Because waiting is slow. This approach can greatly improve performance under light contention IIRC.

Reply:
that makes a lot of sense, do you have an example I could look at so I can understand this design better?

Reply:
I can only remember this [blog post](https://matklad.github.io/2020/01/04/mutexes-are-faster-than-spinlocks.html). But I do have seen this opinion in many places. Just think of many small critical regions contending for a single mutex. In this case, a few spins can avoid many syscalls.

Reply:
BTW, this PR looks pretty much similar to Rust's mutex implementation.

FYI: https://github.com/rust-lang/rust/blob/8387315ab3c26a57a1f53a90f188f0bc88514bca/library/std/src/sys/sync/mutex/futex.rs

Reply:
I left the `spin_count` as a configurable option. One can set `LIBC_CONF_RAW_MUTEX_DEFAULT_SPIN_COUNT` to `0` to disable such behavior.

Reply:
Notice that the spin loops continues only if the lock is hold by someone and contention is not heavy. In that case, the spinning thread knows that there is someone making progress in their critical section and the spinning one will likely obtain the lock once that critical section is over. Therefore, if the critical section is short, considerable syscalls can be avoided.

Notice that `CndVar`, `RwLock` exhibits waiting queues protected by mutexes. The critical section concerning the update of the queue is actually small. Hence, I think it may be a good choice to spin several rounds before going to sleep (futex_wait). 

I currently choose `100` as the default value because it aligns with the choice of Rust's standard library.

Reply:
This does look very similar to the rust implementation, I have two comments about that: First, make sure that you're not just blindly copying what they do, it might not match the use case or semantics of libc. Second, make sure that you're not copying the code because that could have license implications. It should be fine as long as you're primarily writing the code yourself (but also I am not a lawyer).

As for leaving the spinning in, that's fine with me now that I understand the purpose of it.

Reply:
To my understanding, besides spin, Rust's mutex and rwlock implementation itself is largely similar to `bionic`'s implementation (simplified because Rust does not support complicated pthread features). See 
- https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/pthread_mutex.cpp. 
- https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/pthread_rwlock.cpp. 

The code here  is not directly copied from Rust, whose `RawMutex` does not have pshared and timeout support (In fact, I was told that Rust's standard library is not designed to support forking the process without replace the executable image).

However, I do think Rust has good clean and tidy codebase to study. Its platform abstraction layer can be very useful when learning about fundamental runtime operations.

Reply:
taking design inspiration from rust should be fine, and same for bionic. This code doesn't seem like it will be a problem but I wanted to make sure you were aware.

Reply:
heh, I was just poking through the book Rust Atomics and Lock: Low-Level Concurrency in Practice by Mara Bos for inspiration, coincidentally.